pkg/pushsync/pushsync.go

```diff
@@ -310,84 +299,87 @@ func (ps *PushSync) pushToClosest(ctx context.Context, ch swarm.Chunk) (rr *pb.R
-
-		streamer, err := ps.streamer.NewStream(ctx, peer, nil, protocolName, protocolVersion, streamName)
-		if err != nil {
-			lastErr = fmt.Errorf("new stream for peer %s: %w", peer.String(), err)
-			continue
-		}
-		deferFuncs = append(deferFuncs, func() { go streamer.FullClose() })
-
-		// Reserve to see whether we can make the request
-		err = ps.accounting.Reserve(ctx, peer, receiptPrice)
-		if err != nil {
-			return nil, fmt.Errorf("reserve balance for peer %s: %w", peer.String(), err)
-		}
-		deferFuncs = append(deferFuncs, func() { ps.accounting.Release(peer, receiptPrice) })
-
-		w, r := protobuf.NewWriterAndReader(streamer)
-		ctxd, canceld := context.WithTimeout(ctx, timeToLive)
-		deferFuncs = append(deferFuncs, func() { canceld() })
-		if err := w.WriteMsgWithContext(ctxd, &pb.Delivery{
-			Address: ch.Address().Bytes(),
-			Data:    ch.Data(),
-		}); err != nil {
-			_ = streamer.Reset()
-			lastErr = fmt.Errorf("chunk %s deliver to peer %s: %w", ch.Address().String(), peer.String(), err)
-			continue
-		}
-
-		ps.metrics.TotalSent.Inc()
+		go func(peer swarm.Address, ch swarm.Chunk) {
+			ctxd, canceld := context.WithTimeout(ctx, timeToLive)
+			defer canceld()
 
-		// if you manage to get a tag, just increment the respective counter
-		t, err := ps.tagger.Get(ch.TagID())
-		if err == nil && t != nil {
-			err = t.Inc(tags.StateSent)
+			r, err := ps.pushPeer(ctxd, peer, ch)
 			if err != nil {
-				lastErr = fmt.Errorf("tag %d increment: %v", ch.TagID(), err)
-				err = lastErr
-				return nil, err
+				logger.Debugf("could not push to peer %s: %v", peer.String(), err)
+				return
+			}
+			select {
+			case c <- r:
+			case <-done:
+			case <-ctx.Done():
 			}
+		}(peer, ch)
+		select {
+		case r := <-c:
+			return r, nil
+		case <-done:
+			return nil, ErrNoPush
+		case <-ctx.Done():
+			return nil, ctx.Err()
 		}
+	}
+	return nil, ErrNoPush
+}
 
-		var receipt pb.Receipt
-		if err := r.ReadMsgWithContext(ctxd, &receipt); err != nil {
-			_ = streamer.Reset()
-			lastErr = fmt.Errorf("chunk %s receive receipt from peer %s: %w", ch.Address().String(), peer.String(), err)
-			continue
-		}
+func (ps *PushSync) pushPeer(ctx context.Context, peer swarm.Address, ch swarm.Chunk) (*pb.Receipt, error) {
+	// compute the price we pay for this receipt and reserve it for the rest of this function
+	receiptPrice := ps.pricer.PeerPrice(peer, ch.Address())
 
-		if !ch.Address().Equal(swarm.NewAddress(receipt.Address)) {
-			// if the receipt is invalid, try to push to the next peer
-			lastErr = fmt.Errorf("invalid receipt. chunk %s, peer %s", ch.Address().String(), peer.String())
-			continue
-		}
+	streamer, err := ps.streamer.NewStream(ctx, peer, nil, protocolName, protocolVersion, streamName)
+	if err != nil {
+		return nil, fmt.Errorf("new stream for peer %s: %w", peer.String(), err)
+	}
+	defer streamer.Close()
+
+	// Reserve to see whether we can make the request
+	err = ps.accounting.Reserve(ctx, peer, receiptPrice)
```

Comment:
Seems like this can fail. It would be better to move this before the `NewStream` call. That way if we fail here we don't waste additional effort.